File: Bouchonnois.Tests/Unit/Tirer.cs
using Bouchonnois.Domain.Tirer;
using Bouchonnois.Tests.Builders;
using Bouchonnois.UseCases;

namespace Bouchonnois.Tests.Unit
{
    public class Tirer : UseCaseTest<UseCases.Tirer, VoidResponse>
    {
        public Tirer() : base((r, p) => new UseCases.Tirer(r, TimeProvider))
        {
        }

        [Fact]
        public async Task AvecUnChasseurAyantDesBalles()
        {
            Given(
                await UnePartieDeChasseExistante(
                    SurUnTerrainRicheEnGalinettes()
                        .Avec(Bernard())
                ));

            When(id => UseCase.Handle(new Domain.Tirer.Tirer(id, Data.Bernard)));


            Then((response, savedPartieDeChasse) =>
            {
                response.Should().Be(VoidResponse.Empty);
                savedPartieDeChasse
                    .Should()
                    .HaveEmittedEvent(Repository, new ChasseurATiré(savedPartieDeChasse!.Id, Now, Data.Bernard))
                    .And
                    .ChasseurATiré(Data.Bernard, ballesRestantes: 7)
                    .And
                    .GalinettesSurLeTerrain(3);
            });
        }

        public class Echoue : UseCaseTest<UseCases.Tirer, VoidResponse>
        {
            public Echoue() : base((r, p) => new UseCases.Tirer(r, TimeProvider))
            {
            }

            [Fact]
            public void CarPartieNexistePasSansException()
            {
                Given(UnePartieDeChasseInexistante());

                When(partieDeChasseId =>
                    UseCase.Handle(new Domain.Tirer.Tirer(partieDeChasseId, Data.Bernard)));

                ThenFailWith(
                    $"La partie de chasse {PartieDeChasseId} n'existe pas",
                    savedPartieDeChasse => savedPartieDeChasse.Should().BeNull()
                );
            }

            [Fact]
            public async Task CarLeChasseurNestPasDansLaPartie()
            {
                Given(
                    await UnePartieDeChasseExistante(
                        SurUnTerrainRicheEnGalinettes()
                    ));

                When(id => UseCase.Handle(new Domain.Tirer.Tirer(id, Data.ChasseurInconnu)));

                ThenFailWith($"Chasseur inconnu Chasseur inconnu",
                    savedPartieDeChasse => savedPartieDeChasse
                        .Should()
                        .HaveEmittedEvent(Repository,
                            new ChasseurInconnuAVouluTiré(savedPartieDeChasse!.Id, Now, Data.ChasseurInconnu))
                );
            }

            [Fact]
            public async Task AvecUnChasseurNayantPlusDeBalles()
            {
                Given(
                    await UnePartieDeChasseExistante(
                        SurUnTerrainRicheEnGalinettes()
                            .Avec(Dédé(), Bernard().SansBalles(), Robert())
                    ));

                When(id => UseCase.Handle(new Domain.Tirer.Tirer(id, Data.Bernard)));

                ThenFailWith("Bernard tire -> T'as plus de balles mon vieux, chasse à la main",
                    savedPartieDeChasse => savedPartieDeChasse
                        .Should()
                        .HaveEmittedEvent(Repository,
                            new ChasseurSansBallesAVouluTiré(savedPartieDeChasse!.Id, Now, Data.Bernard, "tire"))
                );
            }

            [Fact]
            public async Task SiLesChasseursSontEnApero()
            {
                Given(
                    await UnePartieDeChasseExistante(
                        SurUnTerrainRicheEnGalinettes()
                            .ALapéro()
                    ));

                When(id => UseCase.Handle(new Domain.Tirer.Tirer(id, Data.ChasseurInconnu)));

                ThenFailWith("Chasseur inconnu veut tirer -> On tire pas pendant l'apéro, c'est sacré !!!",
                    savedPartieDeChasse =>
                        savedPartieDeChasse
                            .Should()
                            .HaveEmittedEvent(Repository,
                                new ChasseurAVouluTiréPendantLApéro(savedPartieDeChasse!.Id, Now, Data.ChasseurInconnu))
                );
            }

            [Fact]
            public async Task SiLaPartieDeChasseEstTerminée()
            {
                Given(
                    await UnePartieDeChasseExistante(
                        SurUnTerrainRicheEnGalinettes()
                            .Terminée()
                    ));

                When(id => UseCase.Handle(new Domain.Tirer.Tirer(id, Data.ChasseurInconnu)));

                ThenFailWith("Chasseur inconnu veut tirer -> On tire pas quand la partie est terminée",
                    savedPartieDeChasse =>
                        savedPartieDeChasse
                            .Should()
                            .HaveEmittedEvent(Repository,
                                new ChasseurAVouluTiréQuandPartieTerminée(savedPartieDeChasse!.Id, Now,
                                    Data.ChasseurInconnu))
                );
            }
        }
    }
}

File: java/src/test/java/bouchonnois/tests/unit/Tirer.java
package bouchonnois.tests.unit;

import bouchonnois.service.exceptions.*;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import static bouchonnois.tests.assertions.PartieDeChasseAssert.assertPartieDeChasse;
import static bouchonnois.tests.builders.ChasseurBuilder.*;
import static bouchonnois.tests.builders.PartieDeChasseBuilder.surUnTerrainRicheEnGalinettes;
import static bouchonnois.tests.builders.PartieDeChasseBuilder.unePartieDeChasseInexistante;
import static org.assertj.core.api.Assertions.assertThat;

class Tirer extends PartieDeChasseServiceTests {
    @Test
    void avec_un_chasseur_ayant_des_balles_et_assez_de_galinettes_sur_le_terrain() {
        given(
                unePartieDeChasseExistante(
                        surUnTerrainRicheEnGalinettes()
                                .avec(bernard())
                ));

        when(id -> partieDeChasseService.tirer(id, BERNARD));

        then(savedPartieDeChasse -> assertPartieDeChasse(savedPartieDeChasse)
                .hasEmittedEvent(now, "Bernard tire")
                .chasseurATiré(BERNARD, 7)
                .galinettesSurLeTerrain(3)
        );
    }

    @Nested
    class Echoue {
        @Test
        void car_partie_nexiste_pas() {
            given(unePartieDeChasseInexistante());

            when(id -> partieDeChasseService.tirer(id, BERNARD));

            thenThrow(LaPartieDeChasseNexistePas.class, savedPartieDeChasse -> assertThat(savedPartieDeChasse).isNull());
        }

        @Test
        void avec_un_chasseur_nayant_plus_de_balles() {
            given(
                    unePartieDeChasseExistante(
                            surUnTerrainRicheEnGalinettes()
                                    .avec(dédé(), bernard().sansBalles(), robert())
                    ));

            when(id -> partieDeChasseService.tirer(id, BERNARD));

            thenThrow(TasPlusDeBallesMonVieuxChasseALaMain.class, savedPartieDeChasse ->
                    assertPartieDeChasse(savedPartieDeChasse)
                            .hasEmittedEvent(now, "Bernard tire -> T'as plus de balles mon vieux, chasse à la main")
            );
        }

        @Test
        void car_le_chasseur_n_est_pas_dans_la_partie() {
            given(
                    unePartieDeChasseExistante(
                            surUnTerrainRicheEnGalinettes()
                    ));

            when(id -> partieDeChasseService.tirer(id, CHASSEUR_INCONNU));

            thenThrow(ChasseurInconnu.class, savedPartieDeChasse -> assertThat(savedPartieDeChasse).isNull());
        }

        @Test
        void si_les_chasseurs_sont_en_apéro() {
            given(
                    unePartieDeChasseExistante(
                            surUnTerrainRicheEnGalinettes()
                                    .aLapéro()
                    ));

            when(id -> partieDeChasseService.tirer(id, CHASSEUR_INCONNU));

            thenThrow(OnTirePasPendantLapéroCestSacré.class, savedPartieDeChasse ->
                    assertPartieDeChasse(savedPartieDeChasse)
                            .hasEmittedEvent(now, "Chasseur inconnu veut tirer -> On tire pas pendant l'apéro, c'est sacré !!!")
            );
        }

        @Test
        void si_la_partie_est_terminée() {
            given(
                    unePartieDeChasseExistante(
                            surUnTerrainRicheEnGalinettes()
                                    .terminée()
                    ));

            when(id -> partieDeChasseService.tirer(id, CHASSEUR_INCONNU));

            thenThrow(OnTirePasQuandLaPartieEstTerminée.class, savedPartieDeChasse ->
                    assertPartieDeChasse(savedPartieDeChasse)
                            .hasEmittedEvent(now, "Chasseur inconnu veut tirer -> On tire pas quand la partie est terminée")
            );
        }
    }
}